lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,256 @@
+import 'dart:collection';
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+Element _getLeftElement(AssignmentExpression assignment) {
+  final leftPart = assignment.leftHandSide;
+  return leftPart is SimpleIdentifier
+      ? DartTypeUtilities.getCanonicalElement(leftPart.bestElement)
+      : leftPart is PropertyAccess
+          ? DartTypeUtilities
+              .getCanonicalElement(leftPart.propertyName.bestElement)
+          : null;
+}
+
+bool _isBreakStatement(Statement statement) => (statement is BreakStatement ||
+    (statement is Block &&
+        statement.statements.length == 1 &&
+        statement.statements.first is BreakStatement));
+
+bool _isContinueStatement(
+        Statement statement) =>
+    (statement is ContinueStatement ||
+        (statement is Block &&
+            statement.statements.length == 1 &&
+            statement.statements.first is ContinueStatement));
+
+bool _isReturnStatement(Statement statement) => (statement is ReturnStatement ||
+    (statement is Block &&
+        statement.statements.length == 1 &&
+        statement.statements.first is ReturnStatement));
+
+abstract class ConditionScopeVisitor extends RecursiveAstVisitor {
+  final Queue<Queue<_ExpressionBox>> environments = new Queue();
```

Comment:
I would start by writing a class like the following:
```
class ConditionScope {
  /**
   * The outer scope, of `null` if this is the outermost scope.
   */
  final Scope outer;

  /**
   * Initialize a newly created scope to have the given [outer] scope.
   */
  ConditionScope(this.outer);
}
```
Then I would add fields and methods to support adding `_ExpressionBox`s and performing lookup (which would replace one of the method in `ConditionScopeVisitor`). Given that you're not really treating the inner queues as queues, I'd probably change to using a list. The `environments` field would be replaced by
```
ConditionScope currentScope;
```
pushes look like
```
currentScope = new ConditionScope(currentScope);
```
and pops look like
```
currentScope = currentScope.outer;
```